### PR TITLE
Reland double gzip wrapping NOTICES to reduce on-disk installed space

### LINKED ISSUE #71102: Deduplicate NOTICES
On 1.24.0-8.0.pre.344, the `LicenseResult` collected from LicenseCollector https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/asset.dart#L322 produce a lot of redundancies. 

The uncompressed size of the file is relevant for https://github.com/flutter/flutter/issues/71098. 

A breakdown of occurrences per library shows:

```
  86 harfbuzz
  75 boringssl
  47 skia
  34 turbo
  34 libjpeg
  29 angle
  25 zlib
  22 icu
  19 freetype2
  17 dart
  12 fuchsia_sdk
   9 libwebp
   8 glfw
   4 rapidjson
   4 files
   4 base
   3 libcxx
   3 khronos
   3 double
   3 cpp
   3 conversion
   3 abseil
   2 xxhash
   2 tcmalloc
   2 skcms
   2 root_certificates
   2 ffx_spd
   1 wuffs
   1 wasmer
   1 vulkanmemoryallocator
   1 vector_math
   1 test_api
   1 term_glyph
   1 smhasher
   1 pkg
   1 libsdl
   1 flutter
   1 engine
   1 colorama
   1 clock
   1 charcode
   1 characters
   1 boolean_selector
   1 async
   1 StackWalker
```

Some examples of "different" entries:

```
skia

Copyright 2018 Google Inc.
...
```

```
skia

Copyright 2018 Google LLC
...
```

```
skia

Copyright 2018 Google LLC.
```

```
skia

Copyright 2018 Google, LLC
```

Do we need all these variations?

### PR DESCRIPTION
**Pull request by xster**
Reland the revert at #71899 to fix #71102

The difference between this and the last PR is that it now first compares uncompressed strings against the existing entry first before doing any work to avoid taking a performance hit. 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by xster** [comment]
@jonahwilliams did I get what you were telling me last time right?

File: packages/flutter_tools/lib/src/asset.dart [hunk 1]
```diff
@@ -141,6 +143,13 @@ class ManifestAssetBundle implements AssetBundle {
 
   static const String _kAssetManifestJson = 'AssetManifest.json';
   static const String _kNoticeFile = 'NOTICES';
+  // Comically, this can't be name with the more common .gz file extension
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/asset.dart:146]: oh Gradle

File: packages/flutter_tools/lib/src/asset.dart [hunk 2]
```diff
@@ -425,7 +434,23 @@ class ManifestAssetBundle implements AssetBundle {
...
+      // Don't compress the NOTICES file on web since the client doesn't have
+      // dart:io to decompress it.
+      _setIfChanged(_kNoticeFile, DevFSStringContent(licenseResult.combinedLicenses));
+    } else {
+      if (!entries.containsKey(_kNoticeZippedFile) ||
+          !(entries[_kNoticeZippedFile] as DevFSStringCompressingBytesContent)
+              .equals(licenseResult.combinedLicenses)) {
+        entries[_kNoticeZippedFile] = DevFSStringCompressingBytesContent(
+          licenseResult.combinedLicenses,
+          // A zlib dictionary is a hinting string sequence with the most
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/asset.dart:447]: TIL

File: packages/flutter_tools/lib/src/devfs.dart [hunk 3]
```diff
@@ -203,6 +203,65 @@ class DevFSStringContent extends DevFSByteContent {
...
+class DevFSStringCompressingBytesContent extends DevFSContent {
+  DevFSStringCompressingBytesContent(this._string, { String hintString })
+    : _compressor = ZLibEncoder(
+      dictionary: hintString == null
+          ? null
+          : utf8.encode(hintString),
+      gzip: true,
+      level: 9,
+    ),
+    super();
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/devfs.dart:225]: nit: `super()` isn't needed`

File: packages/flutter_tools/lib/src/asset.dart [hunk 4]
```diff
@@ -425,7 +434,23 @@ class ManifestAssetBundle implements AssetBundle {
 
     _setIfChanged(_kAssetManifestJson, assetManifest);
     _setIfChanged(kFontManifestJson, fontManifest);
-    _setIfChanged(_kNoticeFile, licenses);
+    if (targetPlatform == TargetPlatform.web_javascript) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/asset.dart:437]: I would split up this code into a method with good documentation on what it is doing. Its a bit hard to follow the conditional as is

File: packages/flutter_tools/lib/src/devfs.dart [hunk 5]
```diff
@@ -203,6 +203,65 @@ class DevFSStringContent extends DevFSByteContent {
...
+
+  final String _string;
+  final ZLibEncoder _compressor;
+  final DateTime _modificationTime = DateTime.now();
+
+  List<int> _bytes;
+  bool _isModified = true;
+
+  List<int> get bytes {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/devfs.dart:234]: `List<int> get bytes => _bytes ??= _compressor.convert(utf8.encode(_string));`

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/devfs.dart:234]: This will have fewer issues once null safety is on for the file, since the _bytes null check won't promote

File: packages/flutter_tools/lib/src/devfs.dart [hunk 6]
```diff
@@ -203,6 +203,65 @@ class DevFSStringContent extends DevFSByteContent {
...
+  }
+
+  @override
+  int get size => bytes.length;
+
+  @override
+  Future<List<int>> contentsAsBytes() async => bytes;
+
+  @override
+  Stream<List<int>> contentsAsStream() =>
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/devfs.dart:259]: I think there is a `Stream.value` or similar you could use here

**Comment by jonahwilliams** [comment]
Seems to do the job, thank you @xster

File: packages/flutter_tools/lib/src/devfs.dart (hunk 3 above)
Comment (xster) [reply packages/flutter_tools/lib/src/devfs.dart]: ah man, I haven't written dart for so long 😔

File: packages/flutter_tools/lib/src/asset.dart [hunk 7]
```diff
@@ -82,6 +83,7 @@ abstract class AssetBundle {
     String assetDirPath,
     @required String packagesPath,
     bool deferredComponentsEnabled = false,
+    TargetPlatform targetPlatform,
```
Comment (xster) [inline packages/flutter_tools/lib/src/asset.dart:86]: @chingjun I unfortunately needed to add another parameter to AssetBundle.build. I took a quick look in cs and it doesn't seem like there's any AssetBundle subclasses that implements build.

**Comment by xster** [comment]
Note: I took the unzipping part out of the framework's AssetBundle and just put it into the service binding's implementation to reduce the API break surface area.

**Review by jonahwilliams — APPROVED** [review]
Tool changes LGTM

**Comment by xster** [comment]
Not sure why my frob cls were deleted. But it's also expected to fail. I'll submit a g3fix
